### PR TITLE
Added store hours and store details

### PR DESCRIPTION
**Pull request by thumn**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
- Store cards now display whether the store is currently open or what time it will next be open
- The store card now navigates to a separate store card on click that provides information on the store's address, phone number, hours by day of the week, and accepted programs (NOTE: accepted programs is not finished)
- Detailed store hour schedules bold the current day and hours

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

## Relevant Links
N/A

### Online sources
[Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)

[//]: # "Optional - copy links to any tutorial or documentation that was useful to you when working on this PR"

### Related PRs
N/A

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## How to review
Navigate through stores and make sure each store's card can successfully navigate to its corresponding store screen. The current day should be correctly bolded, and the store open status summary should correctly reflect whether the store is open or when it'll next be open. For example, Yang Market has a particular complicated store hours input.

[//]: # "The order in which to review files and what to expect when testing locally"

## Next steps
Start work on tags for accepted programs in the stores list. This will also be applied to the store screens.

NOTE: I'm getting a "each child in a list should have a unique 'key' prop" error -- does anyone know how to fix this? The file is `StoreHours.js`. I also can't figure out how to make the `<StoreDetailText>` primary green on the `StoreCard`, but it works fine on the store screen. (Might be passing in the wrong prop, but the component seems to take in `color` as a prop so I'm not too sure.)

Also, I sometimes get a weird "store is undefined" error on `MapScreen` that goes away if I reload the app. Sadly this might end up circling back to the `useFocusEffect` hook issue/refactoring task we set aside :(

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

## Tests Performed, Edge Cases
Tested with hard-coded store hour inputs:
`9am-8pm Tu-Sat, 9am-7pm Sun, Closed Mon`
`8am-7pm Daily`
`Open 24/7`
`7am-Midnight Daily`

Also manually compared hours between Stores table in the Airtable and the ones that showed up in the store screens. I looked up the more edge-case-y stores and compared the info on the store card and corresponding store screen. (e.g. Yang Market, Mellon Market, T & G Grocery, Smiley Mart).

[//]: # "Hopefully we will add a testing suite/CI soon, but until then note down the steps you took to test locally"

### Screenshots
IMG_2708
IMG_2709

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @tommypoa @wangannie

[//]: # "This tags in both Annies as a default. Feel free to change, or add on anyone who you should be in on the conversation."

### CODE REVIEW

**Comment by thumn** [comment]
LOL Just remembered I'm using the wrong-facing phone icon...

File: screens/map/StoreScreen.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          {/* Directions */}
+          <TouchableOpacity style={{ paddingBottom: 32 }}>
+            <InLineContainer style={{ alignItems: 'center' }}>
+              <FontAwesome5 name="directions" size={24} color={Colors.black} />
+              <Body style={{ marginLeft: 12 }}>{store.address}</Body>
+            </InLineContainer>
+          </TouchableOpacity>
+          {/* Phone Number */}
+          <InLineContainer style={{ alignItems: 'center', paddingBottom: 32 }}>
+            <FontAwesome5 name="phone" solid size={24} color={Colors.black} />
```
Comment (wangannie) [inline screens/map/StoreScreen.js:45]: Change all instances of `Colors.black` to `activeText`

File: components/store/StoreHours.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  const today = new Date();
+  const todaysDayIndex = today.getDay();
+  const todaysDay = daysOfTheWeekFull[todaysDayIndex];
+
+  const hoursList = dictHours => {
+    return (
+      <View style={{ display: 'flex', flexDirection: 'row' }}>
+        {/* Days of the week */}
+        <View style={{ width: '50%' }}>
```
Comment (wangannie) [inline components/store/StoreHours.js:35]: Nit: Maybe adjust this 50-50 ratio to more 40-60 just to make sure we don't get weird wrapping/truncating edge cases on narrower screens.

File: components/store/StoreHours.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          {daysOfTheWeekFull.map((day, _) => {
+            if (day === todaysDay) {
+              return <TabSelected>{day}</TabSelected>;
+            }
+            return <Body color={Colors.secondaryText}>{day}</Body>;
+          })}
+        </View>
+        <View style={{ width: '50%' }}>
+          {daysOfTheWeekFull.map((day, _) => {
+            if (day == todaysDay) {
```
Comment (wangannie) [inline components/store/StoreHours.js:45]: I think we should make sure all of these are `===` instead of `==` bc eslint haha

File: lib/mapUtils.js [hunk 4]
```diff
@@ -43,3 +43,156 @@ export async function getProductData(store) {
 export const getMaxWidth = function sync(screenWidth, ebt, seeProducts) {
   return screenWidth - 32 - 12 + (ebt ? -48 : 0) + (seeProducts ? -130 : 0);
 };
+
+export function constructHoursDict(hours) {
+  const daysOfTheWeek = {
```
Comment (wangannie) [inline lib/mapUtils.js:48]: Wondering if we should consolidate some of these things into a file in constants? Lots of repeated code between these different functions.

File: components/store/StoreHours.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            return <Body color={Colors.secondaryText}>{dictHours[day]}</Body>;
+          })}
+        </View>
+      </View>
+    );
+  };
+
+  const parseHours = () => {
+    // Case: hours = "Open 24/7"
+    if (hours === 'Open 24/7' || hours === 'Store hours unavailable') {
```
[the remark this answers is not in the document]
Comment (wangannie) [reply components/store/StoreHours.js:57]: Actually, can you make it catch any parsing errors and display this 'Store hours unavailable' and log the error just so the whole app doesn't crash whenever there's a tiny mistake in the input?

File: lib/mapUtils.js [hunk 6]
```diff
@@ -43,3 +43,156 @@ export async function getProductData(store) {
...
+    Sat: 'Saturday',
+  };
+  const daysOfTheWeekAbbrev = Object.keys(daysOfTheWeek);
+  const daysOfTheWeekFull = Object.values(daysOfTheWeek);
+  const hoursStrSplit = hours.split(' ');
+
+  // Case: Hours are daily if there are only two strings
+  // e.g. 8am-7pm Daily, 7am-Midnight Daily
+
+  // We expect hoursPerDayDict to look like {'Sunday': '9am-9pm', 'Monday: '7am-12am', ...}
```
[the remark this answers is not in the document]
Comment (wangannie) [reply lib/mapUtils.js:64]: OH also since we changed the input format Midnight to 12am, could we switch that back to display 'Midnight' if it is 12am? Could probably be done with a `.replace` or something....just bc I 12am/midnight/noon always confuses me and I understand why they originally wanted it to say Midnight 😅

File: lib/mapUtils.js [hunk 7]
```diff
@@ -43,3 +43,156 @@ export async function getProductData(store) {
...
+  const openingMin = openingHourMinSplit[1];
+
+  const closingSuffix = closingTime.slice(-2);
+  const closingHourMin = closingTime.slice(0, -2);
+  const closingHourMinSplit = closingHourMin.split(':'); // e.g. ["8", "30"]
+  let closingHour = closingHourMinSplit[0];
+  const closingMin = closingHourMinSplit[1];
+
+  // Convert to 24 hour time
+  if (openingSuffix === 'pm') {
```
[the remark this answers is not in the document]
Comment (thumn) [reply lib/mapUtils.js:168]: In the airtable it looked like the hours were inputted in the form of:
`7:30am-10pm Daily`
`7am-10pm Mon-Sat, 9am-9pm Sun`

Would it be safe for me to assume that it would follow the `timeInterval Day` format (regardless of weird spacing issues)? I haven't accounted for the ones you linked in the picture

File: components/store/StoreHours.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (thumn) [reply components/store/StoreHours.js:57]: Yea I do that at the end of `parseHours`!

[the remark this answers is not in the document]
Comment (thumn) [reply components/store/StoreHours.js:57]: Also fixed to catch null/empty hours!

File: lib/mapUtils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (thumn) [reply lib/mapUtils.js:64]: Wait so if they input `7am-12am` we want the app to show `7am-Midnight` instead?

File: lib/mapUtils.js [hunk 8]
```diff
@@ -43,3 +43,156 @@ export async function getProductData(store) {
...
+    openingHour += 12;
+  }
+
+  if (closingSuffix === 'pm') {
+    closingHour += 12;
+  }
+
+  const convertedOpeningTime = openingHour * 60 + openingMin;
+  const convertedClosingTime = closingHour * 60 + closingMin;
```
[the remark this answers is not in the document]
Comment (thumn) [reply lib/mapUtils.js:177]: Ah shoot I forgot the minutes were strings RIP thanks for catching that

[the remark this answers is not in the document]
Comment (thumn) [reply lib/mapUtils.js:177]: Ahh also the issue was that some of the times don't have minutes so I didn't check for minutes before converting something that didn't exist! So that's why it's NaN but that's fixed now tyty

File: lib/mapUtils.js [hunk 9]
```diff
@@ -43,3 +48,178 @@ export async function getProductData(store) {
...
+// Takes in the index representing today's day of the week
+// and a dictionary mapping days to store hours. Returns a string
+// indicating when the store will next be open.
+function computeNextOpenDay(todaysDayIndex, hoursPerDayDict) {
+  // The store is closed, so we need the next day and its opening time
+  let tmrwDayIndex = (todaysDayIndex + 1) % 7;
+  let tmrwDay = daysOfTheWeekFull[tmrwDayIndex];
+  let tmrwStoreHours = hoursPerDayDict[tmrwDay];
+  // Loop until next open day is found.
+  while (tmrwStoreHours === 'Closed') {
```
Comment (wangannie) [inline lib/mapUtils.js:147]: NICE that you catch this! I was expecting this edge case `9am-7pm Sun, 9am-8pm Tu-Thurs, Closed Fri, Closed Sat` to break but that's very cool that it works! Still, I'd probably rename this to be more specific like next opening day bc 'tmrw' is a bit deceiving.

File: components/store/StoreCard.js [hunk 10]
```diff
@@ -55,89 +64,108 @@ export default function StoreCard({
     });
   };
 
+  const storeOpenStatus = computeStoreOpen(store.storeHours);
```
Comment (wangannie) [inline components/store/StoreCard.js:67]: I'm getting REALLY SLOW loading! Is there a way we can make this only call once? It seems to be called with every movement on the map and gets REALLY REALLY slow on the store list screen. When I type to search, it's so slow that it doesn't even respond at all. It gets to a point where I can't load all the stores on storelist and then it fully stops responding altogether like I can't tap map icons or close the Find a Store page.

File: components/store/StoreCard.js [hunk 11]
```diff
@@ -55,89 +64,108 @@ export default function StoreCard({
...
-          )}
-        </RowContainer>
-        {seeProduct && <StoreProductButton callBack={callBack} />}
-      </SpaceBetweenRowContainer>
-      {seeDistance && (
-        <Caption style={{ marginBottom: 4 }} color={Colors.secondaryText}>
-          {`${distance} miles away`}
-        </Caption>
-      )}
+    <TouchableHighlight
```
Comment (wangannie) [inline components/store/StoreCard.js:70]: Ok sorry this totally my bad this was a bad suggestion. This turns black on clicking 😅 TouchableOpacity would work just fine instead since it's on StoreCard and not transparent to the whole mapscreen now 😂

File: lib/mapUtils.js [hunk 12]
```diff
@@ -155,7 +152,8 @@ function computeNextOpenDay(todaysDayIndex, hoursPerDayDict) {
   return 'Closed until '
     .concat(tmrwDay)
     .concat(' ')
-    .concat(tmrwOpeningTime);
+    .concat(tmrwOpeningTime)
+    .replace('12am', 'Midnight');
```
Comment (wangannie) [inline lib/mapUtils.js:156]: I'm p sure you actually wouldn't need this here or in 'Open now until' below bc doing this in the final `StoreHours` does the final change.

Comment (thumn) [reply lib/mapUtils.js:156]: Oh I added it there because I don't display `StoreHours` on the `StoreCard` (this method primarily gives you the summary, aka "Open 24/7" or whatever, and `StoreHours` gives you the whole schedule)

File: components/store/StoreCard.js (hunk 10 above)
Comment (thumn) [reply components/store/StoreCard.js:67]: Yeah good point! I removed one call in the `StoreDetailsScreen` and pass the summary line in as a prop from `StoreCard`. Hopefully this helps!

File: lib/mapUtils.js [hunk 13]
```diff
@@ -43,3 +48,193 @@ export async function getProductData(store) {
...
+  // Case: Hours are daily if there are only two strings
+  // e.g. 8am-7pm Daily, 7am-12am Daily
+
+  // We expect hoursPerDayDict to look like {'Sunday': '9am-9pm', 'Monday': '7am-12am', ...}
+  const hoursPerDayDict = {};
+
+  if (hoursStrSplit.length === 2 && hoursStrSplit.includes('Daily')) {
+    const time = hoursStrSplit[0];
+
+    daysOfTheWeekFull.map((day, _) => {
```
[the remark this answers is not in the document]
Comment (thumn) [reply lib/mapUtils.js:74]: I added keys to all the component children of the loop in `StoreHours` and it seems to get rid of the warning so we good!

File: lib/mapUtils.js [hunk 14]
```diff
@@ -43,3 +49,196 @@ export async function getProductData(store) {
...
+        .concat(nextDay)
+        .concat(' ')
+        .concat(nextOpeningTime)
+        .replace('12am', 'Midnight');
+    }
+    nextDayIndex = (nextDayIndex + 1) % 7;
+    nextDay = daysOfTheWeekFull[nextDayIndex];
+    nextDayStoreHours = hoursPerDayDict[nextDay];
+  }
+  return 'Store hours unavailable lol';
```
Comment (wangannie) [inline lib/mapUtils.js:161]: HAHA don't forget to change this

File: lib/mapUtils.js [hunk 15]
```diff
@@ -43,3 +49,196 @@ export async function getProductData(store) {
...
+  for (let i = 0; i < 7; i++) {
+    if (nextDayStoreHours !== 'Closed') {
+      const nextDayHoursSplit = nextDayStoreHours.split('-');
+      const nextOpeningTime = nextDayHoursSplit[0]; // e.g. "8:30am"
+
+      return 'Closed until '
+        .concat(nextDay)
+        .concat(' ')
+        .concat(nextOpeningTime)
+        .replace('12am', 'Midnight');
```
Comment (kwfk) [inline lib/mapUtils.js:155]: pretty minor but for code readability, it's better to use string interpolation:
```javascript
return `Closed until ${nextDay} ${nextOpeningTime}`.replace('12am', 'Midnight');
```

File: lib/mapUtils.js [hunk 16]
```diff
@@ -43,3 +49,196 @@ export async function getProductData(store) {
...
+
+    const convertedOpeningTime = openingHour * 60 + openingMin;
+    const convertedClosingTime = closingHour * 60 + closingMin;
+
+    // The store is still open today
+    if (
+      convertedOpeningTime <= currentTime &&
+      currentTime <= convertedClosingTime
+    ) {
+      return 'Open now until '.concat(closingTime).replace('12am', 'Midnight');
```
Comment (kwfk) [inline lib/mapUtils.js:235]: same here for code readability:
```javascript
return `Open now until ${closingTime}`.replace('12am', 'Midnight');

File: lib/mapUtils.js (hunk 14 above)
Comment (thumn) [reply lib/mapUtils.js:161]: oOOPs

File: lib/mapUtils.js (hunk 15 above)
Comment (thumn) [reply lib/mapUtils.js:155]: good point ty fred!!

**Comment by wangannie** [comment]
Getting this in the StoreList sometimes
```
VirtualizedList: You have a large list that is slow to update - make sure your renderItem function renders components that follow React performance best practices like PureComponent, shouldComponentUpdate, etc. Object {
  "contentLength": 4548,
  "dt": 754,
  "prevDt": 662,
}
